src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,240 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency"""
+
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar
+
+import numpy as np
+
+T = TypeVar("T")
+
+def RingBufferArray(size: int):
+    return RingBuffer[T](np.empty(shape=(size,), dtype=np.float64))
+
+
+
+class RingBuffer(Generic[T]):
+    def __init__(self, container: Any):
+        """Initialize the ring buffer with the given container.
+
+        Args:
+            container: Container to store the data in.
+        """
+        self.container = container
+        self.write_index = 0
+        self.read_index = 0
+        self.size = 0
```

Comment:
Ideally all internal representation should be private, so it also doesn't appear in the docs.

```suggestion
        self._container = container
        self._write_index = 0
        self._read_index = 0
        self._size = 0
```

I would also rename `_size` to `_len` as IMHO size is a bit ambiguous, it is capacity? it is size in bytes?